src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.h

```diff
@@ -0,0 +1,293 @@
+///
+/// GOTO instructions are tagged as follows:
+///
+/// ```c
+/// foo /* foo */
+///    CALL do_something()                   // loop_id:3 tags:{top-level}
+///    SKIP; // prehead                      // loop_id:3 tags:{top-level}
+/// 1: DECL __not_done1 : signedbv[32]       // loop_id:0 tags:{head}
+///    CALL __not_done1 := not_done1()       // loop_id:0 tags:{body}
+///    IF ¬(__not_done1 ≠ 0) THEN GOTO 7     // loop_id:0 tags:{body,exiting}
+///    CALL do_something1()                  // loop_id:0 tags:{body}
+///    SKIP; // prehead                      // loop_id:0 tags:{body}
+/// 2: DECL __not_done2 : signedbv[32]       // loop_id:1 tags:{head}
+///    CALL __not_done2 := not_done2()       // loop_id:1 tags:{body}
+///    IF ¬(__not_done2 ≠ 0) THEN GOTO 3     // loop_id:1 tags:{body,exiting}
+///    CALL do_something2()                  // loop_id:1 tags:{body}
+///    GOTO 2                                // loop_id:1 tags:{latch}
+/// 3: SKIP                                  // loop_id:0 tags:{body}
+///    DECL __must_break1 : signedbv[32]     // loop_id:0 tags:{body}
+///    CALL __must_break1 := must_break1()   // loop_id:0 tags:{body}
+///    IF ¬(__must_break1 ≠ 0) THEN GOTO 6   // loop_id:0 tags:{body}
+///    SKIP // prehead                       // loop_id:0 tags:{body}
+/// 4: DECL __not_done3 : signedbv[32]       // loop_id:2 tags:{head}
+///    CALL __not_done3 := not_done3()       // loop_id:2 tags:{body}
+///    IF ¬(__not_done3 ≠ 0) THEN GOTO 5     // loop_id:2 tags:{body,exiting}
+///    CALL do_something3()                  // loop_id:2 tags:{body}
+///    GOTO 4                                // loop_id:2 tags:{latch}
+/// 5: SKIP                                  // loop_id:0 tags:{body}
+///    GOTO 7                                // loop_id:0 tags:{body,exiting}
+/// 6: SKIP                                  // loop_id:0 tags:{body}
+///    DEAD __must_break1                    // loop_id:0 tags:{body}
+///    DEAD __not_done2                      // loop_id:0 tags:{body}
+///    GOTO 1                                // loop_id:0 tags:{latch}
+/// 7: SKIP                                  // loop_id:3 tags:{top-level}
+///    DEAD __not_done1                      // loop_id:3 tags:{top-level}
+///    END_FUNCTION                          // loop_id:3 tags:{top-level}
+/// ```
+///
+/// The tags allow to the dynamic frames instrumentation pass to select the
+/// write set instance of the specific loop to instrument the instruction,
+/// and allow the loop contracts instrumentation pass to robustly locate head,
+/// exiting nodes and latch nodes when applying the loop contract
+/// transformation.
+class dfcc_cfg_infot
+{
+public:
+  dfcc_cfg_infot(
+    const irep_idt &function_id,
+    goto_functiont &goto_function,
+    const exprt &top_level_write_set,
+    const dfcc_loop_contract_modet loop_contract_mode,
+    goto_modelt &goto_model,
+    message_handlert &message_handler,
+    dfcc_utilst &utils,
```

Comment:
I do sincerely hope that `dfcc_utilst` is not stateful. I think you should be constructing an instance of it from your `goto_model` and `message_handler` parameters.

Reply:
FWIW: I am working on cleaning up `dfcc_utilst` so as just to be made up of `static` methods. See #7680, #7681, #7682 with more to come.

Reply:
Completed once #7683 is merged.

Reply:
Are these changes breaking #7541 ?

Reply:
Just to also note this down in here: #7683 will not be merged before this PR.